app/services/user_registration_service.rb

```diff
@@ -0,0 +1,20 @@
+class UserRegistrationService < BaseService
+  attr_reader :username, :password, :email
+  def initialize(username, password, email)
+    @username = username
+    @password = password
+    @email = email
+  end
+  def call
```

Comment:
The `UserRegistrationService` class is using manual validation which can be error-prone and hard to maintain. Consider using ActiveModel validations for better maintainability.